File: Android/ExpenseApp/app/src/main/java/com/example/expenseapp/firebase/FirestoreManager.kt
package com.example.expenseapp.firebase

import com.example.expenseapp.model.Expense
import com.example.expenseapp.storage.DataStoreManager
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await
import java.util.Date

// MARK: - FirestoreManager
// DataStoreManager를 상속해 기존 ViewModel과 완전 호환
class FirestoreManager(private val userId: String) : DataStoreManager() {

    private val db: FirebaseFirestore = Firebase.firestore
    private val collection get() = db.collection("users")
        .document(userId)
        .collection("expenses")

    // MARK: - 실시간 동기화 Flow (Firestore Snapshot Listener)
    val expensesFlow: Flow<List<Expense>> = callbackFlow {
        val listener: ListenerRegistration = collection
            .orderBy("date", com.google.firebase.firestore.Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error); return@addSnapshotListener
                }
                val expenses = snapshot?.documents
                    ?.mapNotNull { doc ->
                        runCatching {
                            Expense(
                                id       = doc.getString("id")       ?: doc.id,
                                amount   = (doc.getLong("amount")    ?: 0L).toInt(),
                                category = doc.getString("category") ?: "",
                                note     = doc.getString("note")     ?: "",
                                date     = doc.getDate("date")?.time ?: System.currentTimeMillis()
                            )
                        }.getOrNull()
                    } ?: emptyList()
                trySend(expenses)
            }
        awaitClose { listener.remove() }
    }

    // MARK: - DataStoreManager 구현 (호환용)
    override suspend fun save(expenses: List<Expense>) {
        val batch = db.batch()
        expenses.forEach { expense ->
            val ref = collection.document(expense.id)
            batch.set(ref, expense.toMap())
        }
        batch.commit().await()
    }

    override suspend fun load(): List<Expense> {
        return try {
            collection
                .orderBy("date", com.google.firebase.firestore.Query.Direction.DESCENDING)
                .get().await()
                .documents
                .mapNotNull { doc ->
                    runCatching {
                        Expense(
                            id       = doc.getString("id")       ?: doc.id,
                            amount   = (doc.getLong("amount")    ?: 0L).toInt(),
                            category = doc.getString("category") ?: "",
                            note     = doc.getString("note")     ?: "",
                            date     = doc.getDate("date")?.time ?: System.currentTimeMillis()
                        )
                    }.getOrNull()
                }
        } catch (e: Exception) {
            emptyList()
        }
    }

    // MARK: - 개별 CRUD (권장)

    suspend fun addExpense(expense: Expense): Result<Unit> = runCatching {
        collection.document(expense.id).set(expense.toMap()).await()
    }

    suspend fun updateExpense(expense: Expense): Result<Unit> = runCatching {
        collection.document(expense.id).set(expense.toMap()).await()
    }

    suspend fun deleteExpense(expense: Expense): Result<Unit> = runCatching {
        collection.document(expense.id).delete().await()
    }

    // MARK: - Expense → Firestore Map
    private fun Expense.toMap(): Map<String, Any> = mapOf(
        "id"       to id,
        "amount"   to amount,
        "category" to category,
        "note"     to note,
        "date"     to Date(date)     // Firestore Timestamp로 자동 변환
    )
}

File: iOS/ExpenseApp/ExpenseApp/Firebase/FirestoreManager.swift
import Foundation
import FirebaseFirestore

// MARK: - FirestoreManager
// ExpenseStorageProtocol을 채택하여 기존 ViewModel과 완전 호환
final class FirestoreManager: ExpenseStorageProtocol {

    private let db:     Firestore
    private let userId: String
    private var listener: ListenerRegistration?

    // 실시간 스트림 콜백
    var onExpensesUpdated: (([Expense]) -> Void)?

    private var collection: CollectionReference {
        db.collection("users").document(userId).collection("expenses")
    }

    init(userId: String) {
        self.db     = Firestore.firestore()
        self.userId = userId
        startListening()
    }

    deinit {
        listener?.remove()
    }

    // MARK: - 실시간 동기화 (Firestore Listener)
    private func startListening() {
        listener = collection
            .order(by: "date", descending: true)
            .addSnapshotListener { [weak self] snapshot, error in
                guard let self, let snapshot else {
                    print("Firestore 리스너 오류:", error?.localizedDescription ?? "")
                    return
                }
                let expenses = snapshot.documents.compactMap {
                    try? $0.data(as: FirestoreExpense.self)
                }.map(\.toExpense)

                Task { @MainActor in
                    self.onExpensesUpdated?(expenses)
                }
            }
    }

    // MARK: - ExpenseStorageProtocol 구현

    /// 전체 목록 저장 (로컬→클라우드 Batch Write)
    func save(_ expenses: [Expense]) {
        let batch = db.batch()
        // 기존 문서 삭제 후 새로 쓰는 방식 대신,
        // 개별 CRUD 메서드를 사용하는 것을 권장.
        // 이 메서드는 초기 마이그레이션 용도
        expenses.forEach { expense in
            let ref = collection.document(expense.id.uuidString)
            try? batch.setData(from: FirestoreExpense(expense), forDocument: ref)
        }
        batch.commit { error in
            if let error { print("Batch 저장 오류:", error) }
        }
    }

    /// 초기 로드 (한 번 읽기)
    func load() -> [Expense] {
        // Firestore는 비동기이므로, 초기값은 빈 배열 반환 후
        // startListening()의 onExpensesUpdated 콜백으로 업데이트됨
        return []
    }

    // MARK: - 개별 CRUD (권장)

    func addExpense(_ expense: Expense) async throws {
        let ref = collection.document(expense.id.uuidString)
        try ref.setData(from: FirestoreExpense(expense))
    }

    func updateExpense(_ expense: Expense) async throws {
        let ref = collection.document(expense.id.uuidString)
        try ref.setData(from: FirestoreExpense(expense), merge: true)
    }

    func deleteExpense(_ expense: Expense) async throws {
        try await collection.document(expense.id.uuidString).delete()
    }

    func fetchExpenses() async throws -> [Expense] {
        let snapshot = try await collection
            .order(by: "date", descending: true)
            .getDocuments()
        return snapshot.documents
            .compactMap { try? $0.data(as: FirestoreExpense.self) }
            .map(\.toExpense)
    }
}

// MARK: - Firestore DTO (Codable 직렬화용)
struct FirestoreExpense: Codable {
    @DocumentID var documentId: String?
    let id:       String
    let amount:   Int
    let category: String
    let note:     String
    let date:     Date        // Firestore Timestamp ↔ Date 자동 변환

    init(_ expense: Expense) {
        self.id       = expense.id.uuidString
        self.amount   = expense.amount
        self.category = expense.category
        self.note     = expense.note
        self.date     = expense.date
    }

    var toExpense: Expense {
        Expense(
            id:       UUID(uuidString: id) ?? UUID(),
            amount:   amount,
            category: category,
            note:     note,
            date:     date
        )
    }
}
